awx/main/models/mixins.py

```diff
@@ -483,3 +484,48 @@ def get_active_jobs(self):
             raise RuntimeError("Programmer error. Expected _get_active_jobs() to return a QuerySet.")
 
         return [dict(id=t[0], type=mapping[t[1]]) for t in jobs.values_list('id', 'polymorphic_ctype_id')]
+
+
+class WebhookMixin(models.Model):
+    class Meta:
+        abstract = True
+
+    SERVICES = [
+        ('github', "Github"),
+        ('gitlab', "Gitlab"),
+        ('bitbucket', "Bitbucket"),
+    ]
+
+    webhook_service = models.CharField(
+        max_length=16,
+        choices=SERVICES,
+        blank=True
+    )
+    webhook_key = prevent_search(models.CharField(
+        max_length=64,
+        blank=True
+    ))
+    webhook_credential = models.ForeignKey(
+        'Credential',
+        blank=True,
+        null=True,
+        on_delete=models.SET_NULL,
+        related_name='%(class)ss'
+    )
```

Comment:
I'm not sure it's *strictly* necessary to create `generalized` credentials containers *everywhere*. Sometimes we really just need *one* type of credential. I think this is fine here. If we want to end up supporting multiple webhooks per-endpoint then we might feel different but I feel this would have to change in major ways to support that... given that there are so few webhook sources it's probably fine to have a customer support a 1:1 linkage between a webhook source and a job template-ish.

Reply:
@matburt should it be shown in JT/WFJT summary_fields?

Reply:
Yes, I think so. This credential feels less like the catchall `credentials` since this one won't be included in the playbook runs. I'm open to being conviced otherwise?

Reply:
As much as I like the new `.credentials` M2M, the general purpose of that is intended to be "credentials we use for the job run".  I could see the argument for this being a distinct new foreign key.  What did we do for the Insights credential type and its' relationship?

Reply:
> What did we do for the Insights credential type and its' relationship?

That's a foreign key from inventory, `insights_credential`, yes.

Projects are still sporting a field `credential` for their scm credential (from the subclass table), which is most similar to this.

And actually, I think that the null case shouldn't hurt serialization. It will muck hard with copy / start user_capabilities. We will need to return to that before the end of this cycle and make sure it's not doing something crazy we didn't expect.